bindings/py/src/CMakeLists.txt

```diff
@@ -0,0 +1,194 @@
+#   nupic.core.algorithms.pyd
+#   nupic.core.engine.pyd
+#   nupic.core.math.pyd
+#
+# For PyBind11 info See: https://pybind11.readthedocs.io/en/stable/compiling.html
+#######################################################################
+
+cmake_minimum_required(VERSION 3.11)
+project(nupic_core CXX)
+
+set_property(GLOBAL PROPERTY USE_FOLDERS ON)
+set(CMAKE_VERBOSE_MAKEFILE OFF) # toggle for cmake debug
+
+
+#
+# Python support
+#
+set(src_py_algorithms_files
+    bindings/algorithms/algorithm_module.cpp
+    bindings/algorithms/py_Algorithms.cpp
+    bindings/algorithms/py_Cells4.cpp
+    bindings/algorithms/py_HTM.cpp
+    bindings/algorithms/py_SDRClassifier.cpp
+    bindings/algorithms/py_SpatialPooler.cpp
+	)
+	
+set(src_py_engine_files
+    plugin/PyBindRegion.cpp
+    plugin/PyBindRegion.hpp
+    plugin/RegisteredRegionImplPy.hpp
+    bindings/engine/engine_module.cpp
+    bindings/engine/py_Engine.cpp
+    bindings/engine/py_Region.cpp
+    bindings/engine/py_OS.cpp
+    bindings/engine/py_Timer.cpp
+    bindings/engine/py_utils.hpp
+	)
+	
+set(src_py_math_files
+    bindings/math/math_module.cpp    
+    bindings/math/py_ArrayAlgo.cpp
+    bindings/math/py_Domain.cpp
+    bindings/math/py_Gaussian_2D.cpp
+    bindings/math/py_Math.cpp
+    bindings/math/py_NearestNeighbor.cpp
+    bindings/math/py_Random.cpp
+    bindings/math/py_reals.cpp
+    bindings/math/py_SegmentSparseMatrix.cpp
+    bindings/math/py_Set.cpp
+    bindings/math/py_SparseBinaryMatrix.cpp
+    bindings/math/py_SparseMatrix.cpp
+    bindings/math/py_SparseMatrixAlgorithms.cpp
+    bindings/math/py_SparseMatrixConnections.cpp
+    bindings/math/py_SparseRLEMatrix.cpp
+    bindings/math/py_SparseTensor.cpp
+    bindings/math/PyBindSparseTensor.hpp
+    bindings/math/SparseTensorIndex.hpp
+    )
+    
+set(src_py_test_files
+	tests/main.cpp
+	tests/PyRegionTest.cpp
+	)
+	
+#set up file tabs in Visual Studio
+source_group("algorithms" FILES ${src_py_algorithms_files})
+source_group("engine" FILES ${src_py_engine_files})
+source_group("math" FILES ${src_py_math_files})
+source_group("test" FILES ${src_py_test_files})
+############################################################
+#
+# Build the three shared binding libraries for the Python Interface
+#
+############################################################
+
+
+# set CPP standard. default for PyBind11 is C++14.  If requested standard not available it will fall back to C++11.
+if (MSVC)
+  set(PYBIND11_CPP_STANDARD /std:c++latest) # Enables MSVC C++17
+endif()
+##########################
+#  Fetch pybind11 from Git
+# This will download pybind11 at configure time, and add_subdirectory it. Then you're ready to call pybind11_add_module.
+# See https://stackoverflow.com/questions/47027741/smart-way-to-cmake-a-project-using-pybind11-by-externalproject-add?rq=1
+include(FetchContent)
+FetchContent_Declare(
+	pybind11
+    FETCHCONTENT_BASE_DIR ${EP_BASE}
+    GIT_REPOSITORY https://github.com/pybind/pybind11
```

Comment:
> FETCHCONTENT_BASE_DIR ${EP_BASE}

I am very much against including third party distributions into our distribution.  It blots our distribution.  The full Boost distribution will not even fit....github rejects it as being to large. But any third party distribution should stand on its own.

The CMake 3.11 is required because I want to use FetchContent.  This is similar to External_Project_Add except it performs at configure time rather than execute time.  It is intended for building prerequisites.  Much cleaner but a bummer that it is not installed by default on Ubuntu.  Will have to think about this.

Reply:
I agree that "out-sourcing" the build of deps is smart, 
but pybind is headers-only. It seems natural to include it directly into the codebase. 
Boost is a prerequisite, and possibly will be obsoleted. What I see as a big disadvantage is re-downloading all deps each configure (after git clean), is that correct? And the inability to compile offline from the repo.